{{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/plugin.py

```diff
@@ -118,3 +165,68 @@
 ):
     with open(path, encoding="utf-8") as patch_file:
         hooks.Filters.ENV_PATCHES.add_item((os.path.basename(path), patch_file.read()))
+
+
+########################################
+# CUSTOM JOBS (a.k.a. "do-commands")
+########################################
+
+# A job is a set of tasks, each of which run inside a certain container.
+# Jobs are invoked using the `do` command, for example: `tutor local do importdemocourse`.
+# A few jobs are built in to Tutor, such as `init` and `createuser`.
+# You can also add your own custom jobs:
+
+# To add a custom job, define a Click command that returns a list of tasks,
+# where each task is a pair in the form ("<service>", "<shell_command>").
+# For example:
+@click.command()
+@click.option("-n", "--name", default="plugin developer")
+def say_hi(name: str) -> list[tuple[str, str]]:
+    """
+    An example job that just prints 'hello' from within both LMS and CMS.
+    """
+    return [
+        ("lms", f"echo 'Hello from LMS, {name}!'"),
+        ("cms", f"echo 'Hello from CMS, {name}!'"),
+    ]
+
+
+# Then, add the command function to CLI_DO_COMMANDS:
+hooks.Filters.CLI_DO_COMMANDS.add_item(say_hi)
```

Comment:
Do we want to leave this uncommented by default? I suspect that most people will not take the time to delete this command and will end up with a "say_hi" job in every one of their plugins. (As a sidenote, these identical jobs _will_ generate conflicts...)

Same comment below for the custom CLI command. (which will not conflict but I think that most plugin developers don't need extra custom commands)

Reply:
> I suspect that most people will not take the time to delete this command and will end up with a "say_hi" job in every one of their plugins. (As a sidenote, these identical jobs will generate conflicts...)

Yeah, I hear you.

My concern with the current setup is that there's no syntax checking being done on any of the lines that are commented out. I try to be diligent by manually checking the syntax of the commented-out lines whenever I change them, but [one error](https://github.com/overhangio/cookiecutter-tutor-plugin/pull/16) already slipped past me. Further, it's not just syntax correctness I want to check for: I want the cookie-cut plugin to pass linting, black and mypy [so that we can turn those on by default for new plugins](https://github.com/overhangio/cookiecutter-tutor-plugin/issues/7).

There's another way I could do it, though. I could comment out every line containing code with two `#`s:
```python
# To add a custom job, define a Click command that returns a list of tasks,
# where each task is a pair in the form ("<service>", "<shell_command>").
# For example:
##@click.command()
##@click.option("-n", "--name", default="plugin developer")
##def say_hi(name: str) -> list[tuple[str, str]]:
##    """
##    An example job that just prints 'hello' from within both LMS and CMS.
##    """
##   return [
##        ("lms", f"echo 'Hello from LMS, {name}!'"),
##        ("cms", f"echo 'Hello from CMS, {name}!'"),
##    ]
```

Then, this repository could check the generated plugin.py by substituting `"##"` for `""` and then running pylint/black/mypy on the resulting file.

Reply:
Let me wander in philosophy territory for a minute... I'm perfectly OK with bugs slipping into the source code, as long as we do our best to detect and fix them quickly enough. Two pairs of eyes (yours and mine) are enough to find  issues reported in the forum and fix them. In my experience, trying to preemptively protect ourselves from all sorts of bugs leads to over-complicated processes and engineering.

It's not a black/white thing of course, it's a balance that we need to strike. If you feel like you need the automated syntax checking for every change in the codebase, then sure go for it. As far as I'm concerned, the syntax checking that happens in my IDE is enough for me.

You're the maintainer, so in the end you know what's best. You should just know that no one is going to blame you if there are mistakes in your code. People who don't do anything don't break anything either :)

Reply:
> In my experience, trying to preemptively protect ourselves from all sorts of bugs leads to over-complicated processes and engineering.

Yes, this is something I have found myself susceptible to, and I appreciate you pushing towards simplicity. In retrospect, complicating the *output* of the cookiecutter for the sake of automatic verifiability would have been putting the cart before the horse.

I do feel, though, that keeping the example code in the cookiecutter in a working state is pretty important, since it sits at sensitive point in the Tutor & Open edX adoption funnel: when someone uses it, they might be embarking on the journey to make their first plugin, which I see as a big step towards becoming a Tutor power user and an active community member. If the example code doesn't work, they might move throw their hands up in frustration and move onto something else! (I mean, hopefully, they would ask for help and then our lovely community would help them :) but people are busy, and they might instead just move on to the next thing because they need a quick solution.)

I've pushed another commit, which returns all the code to a commented-out state, but uses a special prefix (`### `) which this repo's Makefile removes when it runs tests. This way, the complexity is relegated to the CI suite of this repo, and shouldn't leak out to the cookiecutter users. Let me know what you think--if it seems wrong, I'm happy to revert it.

Reply:
I appreciate your desire to have correct code, and I think your solution is terrific.